File: src/lib/core/voiceCall/iceRestartElection.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { shouldInitiateIceRestart } from './iceRestartElection';

describe('shouldInitiateIceRestart', () => {
    let warnSpy: ReturnType<typeof vi.spyOn>;

    beforeEach(() => {
        warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
    });

    afterEach(() => {
        warnSpy.mockRestore();
    });

    it('returns true when local pubkey sorts strictly lower than peer', () => {
        expect(
            shouldInitiateIceRestart(
                '0000000000000000000000000000000000000000000000000000000000000001',
                '0000000000000000000000000000000000000000000000000000000000000002'
            )
        ).toBe(true);
        expect(warnSpy).not.toHaveBeenCalled();
    });

    it('returns false when local pubkey sorts strictly higher than peer', () => {
        expect(
            shouldInitiateIceRestart(
                'ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff',
                '0000000000000000000000000000000000000000000000000000000000000001'
            )
        ).toBe(false);
        expect(warnSpy).not.toHaveBeenCalled();
    });

    it('returns false and warns when both pubkeys are byte-equal', () => {
        const hex =
            'abababababababababababababababababababababababababababababababab';
        expect(shouldInitiateIceRestart(hex, hex)).toBe(false);
        expect(warnSpy).toHaveBeenCalledTimes(1);
        expect(warnSpy.mock.calls[0][0]).toContain('IceRestartElection');
    });

    it('returns false and warns when pubkeys differ only in case (still equal after normalization)', () => {
        expect(
            shouldInitiateIceRestart(
                'AAAA000000000000000000000000000000000000000000000000000000000000',
                'aaaa000000000000000000000000000000000000000000000000000000000000'
            )
        ).toBe(false);
        expect(warnSpy).toHaveBeenCalledTimes(1);
    });

    it('normalizes uppercase local input', () => {
        // Local 0xAAAA... > peer 0xBBBB... ? No: after lowercase, "aaaa" < "bbbb".
        expect(
            shouldInitiateIceRestart(
                'AAAA000000000000000000000000000000000000000000000000000000000000',
                'bbbb000000000000000000000000000000000000000000000000000000000000'
            )
        ).toBe(true);
    });

    it('normalizes uppercase peer input', () => {
        // Local "cccc" > peer "BBBB"→"bbbb", so local does NOT initiate.
        expect(
            shouldInitiateIceRestart(
                'cccc000000000000000000000000000000000000000000000000000000000000',
                'BBBB000000000000000000000000000000000000000000000000000000000000'
            )
        ).toBe(false);
    });

    it('handles mixed-case on both sides consistently with all-lowercase', () => {
        const localUpper =
            'AbCdEf0000000000000000000000000000000000000000000000000000000000';
        const peerUpper =
            'aBcDeF0000000000000000000000000000000000000000000000000000000000';
        // Both normalize to the same hex; expect equal-path warning.
        expect(shouldInitiateIceRestart(localUpper, peerUpper)).toBe(false);
        expect(warnSpy).toHaveBeenCalledTimes(1);
    });

    it('lex compare uses string ordering not numeric magnitude', () => {
        // "10" < "9" in string compare. Confirm that even though the
        // numeric interpretation is opposite, the helper follows the
        // string rule (matching the JS sort semantics on Android-side
        // String.compareTo).
        expect(
            shouldInitiateIceRestart(
                '1000000000000000000000000000000000000000000000000000000000000000',
                '9000000000000000000000000000000000000000000000000000000000000000'
            )
        ).toBe(true);
    });
});

File: android/app/src/test/java/com/nospeak/app/IceRestartElectionTest.java
package com.nospeak.app;

import org.junit.Test;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * Pure-Java tests for {@link IceRestartElection}. Mirrors the JS-side
 * tests in {@code src/lib/core/voiceCall/iceRestartElection.test.ts} to
 * keep the two implementations in lock-step. Both implementations MUST
 * agree byte-for-byte on every input so peers running different
 * backends elect the same side.
 */
public class IceRestartElectionTest {

    @Test
    public void returnsTrue_whenLocalSortsStrictlyLower() {
        assertTrue(IceRestartElection.shouldInitiate(
            "0000000000000000000000000000000000000000000000000000000000000001",
            "0000000000000000000000000000000000000000000000000000000000000002"
        ));
    }

    @Test
    public void returnsFalse_whenLocalSortsStrictlyHigher() {
        assertFalse(IceRestartElection.shouldInitiate(
            "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
            "0000000000000000000000000000000000000000000000000000000000000001"
        ));
    }

    @Test
    public void returnsFalse_whenBothByteEqual() {
        String hex = "abababababababababababababababababababababababababababababababab";
        assertFalse(IceRestartElection.shouldInitiate(hex, hex));
    }

    @Test
    public void returnsFalse_whenEqualAfterCaseNormalization() {
        assertFalse(IceRestartElection.shouldInitiate(
            "AAAA000000000000000000000000000000000000000000000000000000000000",
            "aaaa000000000000000000000000000000000000000000000000000000000000"
        ));
    }

    @Test
    public void normalizesUppercaseLocalInput() {
        // "AAAA..." -> "aaaa..." < "bbbb...": local wins.
        assertTrue(IceRestartElection.shouldInitiate(
            "AAAA000000000000000000000000000000000000000000000000000000000000",
            "bbbb000000000000000000000000000000000000000000000000000000000000"
        ));
    }

    @Test
    public void normalizesUppercasePeerInput() {
        // local "cccc" > peer "BBBB" -> "bbbb": local does NOT initiate.
        assertFalse(IceRestartElection.shouldInitiate(
            "cccc000000000000000000000000000000000000000000000000000000000000",
            "BBBB000000000000000000000000000000000000000000000000000000000000"
        ));
    }

    @Test
    public void handlesMixedCaseConsistently() {
        String localUpper = "AbCdEf0000000000000000000000000000000000000000000000000000000000";
        String peerUpper  = "aBcDeF0000000000000000000000000000000000000000000000000000000000";
        // Both normalize to the same hex; expect equal-path return.
        assertFalse(IceRestartElection.shouldInitiate(localUpper, peerUpper));
    }

    @Test
    public void lexCompareUsesStringOrdering_not_numericMagnitude() {
        // "10" < "9" in string compare. Confirm we follow the string
        // rule (matching JS sort semantics on String.compareTo).
        assertTrue(IceRestartElection.shouldInitiate(
            "1000000000000000000000000000000000000000000000000000000000000000",
            "9000000000000000000000000000000000000000000000000000000000000000"
        ));
    }

    @Test
    public void returnsFalse_onNullLocal() {
        assertFalse(IceRestartElection.shouldInitiate(
            null,
            "aaaa000000000000000000000000000000000000000000000000000000000000"
        ));
    }

    @Test
    public void returnsFalse_onNullPeer() {
        assertFalse(IceRestartElection.shouldInitiate(
            "aaaa000000000000000000000000000000000000000000000000000000000000",
            null
        ));
    }

    @Test
    public void returnsFalse_onBothNull() {
        assertFalse(IceRestartElection.shouldInitiate(null, null));
    }
}
